### PR TITLE
Sort JSON-LD data by key

### LINKED ISSUE #362: Consistent order of keys for JSON-LD
It seems almost every time I `jekyll build`, the JSON-LD output appears to have changed when it hasn't.

I use a git-based deployment and it'd be nice to not have every page on the site change for every build to make it easier to review beforehand or afterwards, as well as for size of the repository over time.

I suspect the issue is with the key order not being deterministic. Would you be open to a pull request that makes this deterministic?

cap

### LINKED ISSUE #436: Make JSON-LD output deterministic
Having the same issues reported in #362, where every rendering is appearing as a modified file and triggering unnecessary updates.  Can the output of template.html be made deterministic to support reproducible builds?

If making the output of jsonify deterministic is too difficult, perhaps #135 is a good solution to give users more control where reproducible builds are critical.

### PR DESCRIPTION
**Pull request by parkr**
Deterministic output is desired to improve diffability between builds.
If the output randomly changes order each time it's built, then every
page on a given site will change with every Jekyll build. Tools like
rsync and git which can diff files will always show a change even when
the content hasn't truly changed.

Fixes #362. Fixes #436.

### CODE REVIEW

File: lib/jekyll-seo-tag/drop.rb
Comment (ashmaroli) [inline lib/jekyll-seo-tag/drop.rb:108]: Parsing a JSON string just to sort the Hash object seems sub-optimal to me.
How about sorting the hash at source directly?

Also, `json_ld_sorted_json` sounds repetitive.
How about simply `sorted_json_ld`?

```ruby
def sorted_json_ld
  json_ld.to_h.compact.sort_by(&:first).to_h.to_json
end
```

File: spec/jekyll_seo_tag_integration_spec.rb
Comment (ashmaroli) [inline spec/jekyll_seo_tag_integration_spec.rb:34]: How about using `%()` or outer single-quotes to avoid escaping the JSON's double-quotes?

```suggestion
    expect(json.strip).to eql('{"@context":"https://schema.org","@type":"WebPage","url":"/page.html"}')
```

**Review by ashmaroli — COMMENTED** [review]
I recommend optimizing as much as possible to reduce impact on performance.

File: lib/jekyll-seo-tag/drop.rb
Comment (parkr) [reply lib/jekyll-seo-tag/drop.rb:108]: Since the `to_json` method is doing a bunch of work, it's probably best if I just sort there.

**Comment by parkr** [comment]
@ashmaroli I figured we could make this a first-class thing and just change `to_json` to return a sorted object instead of adding a second method.

**Comment by parkr** [comment]
Ok I decided to measure the `.sort_by(&:first).to_h` I'm adding. It's not great:

```text
Warming up --------------------------------------
            original    55.477k i/100ms
        with sort_by    30.221k i/100ms
Calculating -------------------------------------
            original    539.839k (± 8.8%) i/s -      2.718M in   5.078878s
        with sort_by    322.651k (± 2.8%) i/s -      1.632M in   5.062437s
```

I went up the chain and am taking advantage of the fact that Hashes preserve insertion order. So if you insert "@context" before "@type", then iterating over the Hash (and serializing it to JSON) will give you "@context" first, and "@type" second consistently. I overwrote the original Drop#to_h to sort the keys.

**Comment by ashmaroli** [comment]
> taking advantage of the fact that Hashes preserve insertion order.. overwrote the original Drop#to_h to sort the keys.

This is a wonderful idea! Good job! :clap: 

**I think this is sufficient for an approval now**, but definitely could be optimized further..

Essentially, the `json_ld.to_json` call currently consists of numerous iterations:
```ruby
def to_json(state)
  keys.sort.each_with_object({}) do |(key, _), result|
    result[key] = self[key]
  end.compact.to_json(state)
end
```
:point_right: **`.sort.each_with_object.compact`**

I'll think about this till tomorrow. and shall register my approval then.
Will merge after Matt provides his inputs too.

Thank you very much @parkr

**Comment by parkr** [comment]
@ashmaroli Great point! Further optimized to add an `if <value>` to the `each_with_object`. I think we need to keep `keys.sort` but a good optimization is to get rid of `compact` if not necessary!

**Comment by ashmaroli** [comment]
:+1:

However, this PR ***may have exposed a bug in Jekyll Core***.. Specifically:
```ruby
keys.sort.each_with_object({}) do |(key, _), result|
```
From Jekyll Core codebase [`Drop#keys`](https://github.com/jekyll/jekyll/blob/d46bef774a8f91e7f58d982cccdface7a3d5d26b/lib/jekyll/drops/drop.rb#L192-L196) is **a flat array**:
```ruby
      def keys
        (content_methods |
          mutations.keys |
          fallback_data.keys).flatten
      end
```
Why is the `.each_with_object` block destructuring the **array** into pairs?
```ruby
keys.sort.each_with_object({}) do |(key, _), result|
```
instead of:
```ruby
keys.sort.each_with_object({}) do |key, result|
```

**Comment by parkr** [comment]
Yeah I'm not sure why we did that. I'm going to leave it as-is to match Jekyll Core.

**Review by ashmaroli — COMMENTED** [review]
One final edit and this is good to go.

**Review by ashmaroli — APPROVED** [review]

**Comment by parkr** [comment]
@ashmaroli let's give Matt until the weekend. If he's not free to take a look by then, then I think your and my approval counts as 2 Jekyll maintainers' approval so we can merge and cut a release.

**Review by mattr- — APPROVED** [review]

**Comment by ashmaroli** [comment]
Thanks @parkr
@jekyllbot:  merge +minor

**Comment by parkr** [comment]
Thanks @mattr-! Thanks @ashmaroli!

**Pull request merged**
